### DISCUSSION TITLE
Read multiple server modbus

### DISCUSSION THREAD

**Discussion by Elmatbaisaad** [aldas/modbus-tcp-client#112] [Q&A]
Hello everyone !
I m working in project using symfony and docker and i have three Modbus servers which they have the same ip adress but different ports when i launch the application it connects to the servers because i can define to wich server it should be connected .
the probleme is that i when connect the application to the three servers in the same time it connects but it display 0 as value for the registers and coils.
should I create a container for each server or there is a solution for that ?
Thank you!

**Comment by Elmatbaisaad** [comment]
@aldas

**Reply by aldas** [reply]
I really cant say without seeing some code. 

Requesting multiple different server depends on your requirements. Are those requests limited to some time window? ala application needs to poll once per second etc. Meaning that you can not run things serially. Doing things reliably in parallel is hard and is easier to achieve with dedicated containers for each server (unique ip+port).  My own experience is with long running pollers mainly and for that I created separate cron job for each unique modbus server so problems with latency, plc being offline etc on  one server would not affect results for other polled servers. 

There are examples for ReactPHP and Amp. Both examples request 2 different servers. https://github.com/aldas/modbus-tcp-client/blob/master/examples/example_parallel_requests_reactphp.php

**Reply by Elmatbaisaad** [reply]
Thank you @aldas for your answer.
here my configuration file
![image](https://user-images.githubusercontent.com/101296742/174545758-fbfdb403-f41f-475e-baf9-d71c41fc8d19.png)
i have 3 servers with the same ip adress because they are in the same container but they have different ports

**Reply by aldas** [reply]
What kind of application are you building? is it for long running poller  or some  user interaction driven requests?

**Reply by Elmatbaisaad** [reply]
it's web apllication that will be connected to three real modbus server but now for testing im creating three modbus servers using python and pyModbus library

**Reply by Elmatbaisaad** [reply]
here is the container :
![image](https://user-images.githubusercontent.com/101296742/174546853-211de133-26f7-4081-9ca2-7a94589728eb.png)

**Reply by aldas** [reply]
Does this same setup work as expected when you are requesting from 1 server (remove other 2 from configuration)?  
Also turn on logging as try to see how these 3 server requests and their responses are timed. Are there timeouts? Is response valid modbus packet. If it not timeout and packet is valid then we have problem extracting values from responses.

**Reply by Elmatbaisaad** [reply]
when i try the code locally it works but in docker no its display value 0 for registers and false for coils

**Reply by aldas** [reply]
have you checked what values your modbus server actually responded when you see 0/false? Is is modbus response containing 0x00 as register values?

**Reply by Elmatbaisaad** [reply]
How can i check that because in the code of server , the server send values randomly : 
![image](https://user-images.githubusercontent.com/101296742/174548960-d0f2df4c-4245-4fbc-b740-5fdccd4fd50a.png)

**Reply by aldas** [reply]
You could just print data in PHP side `print_r($response->getData());`  or configure logger https://github.com/aldas/modbus-tcp-client/blob/ac8661a693aed70d542a3c4d10cce08ea1ab88b5/examples/fc3.php#L20

**Reply by Elmatbaisaad** [reply]
yes now i am testing with simple client but in docker it keep send 0

**Reply by aldas** [reply]
so python script is sending 0s? I really cant help with Python parts.

If you look response printed out as hex string then the last bytes are for your requested data. Register filled with 0 would be something like that `0x0000`
https://github.com/aldas/modbus-tcp-client/blob/ac8661a693aed70d542a3c4d10cce08ea1ab88b5/src/Packet/ModbusFunction/ReadHoldingRegistersResponse.php#L20-L27

**Reply by Elmatbaisaad** [reply]
i tried the same code locally it works perfectly but in docker there is a probleme that i can't find it :/

**Reply by aldas** [reply]
are you really sure that this docker block instructs that python script to spawn listener on 3 ports?  For me it seems only to say that map 3 ports from your host to docker container and run some python file. Now that python file that you screenshoted seems to spawn only one server.

**Reply by aldas** [reply]
you could always bash/sh into that python container and see if python is really listening on 3 separate ports.

**Reply by Elmatbaisaad** [reply]
please how can i do that

**Reply by aldas** [reply]
This seems to be good explanation https://geekflare.com/check-docker-network-connections/

**Reply by Elmatbaisaad** [reply]
thaank you so much i will try this
